File: Test-Corpus/claude-combined/B02_synthetic/static-vars-fpts/translated_rust/c_src/src/analyzer.c
/*
 * Copyright 2025 MIT Lincoln Laboratory
 * Permission is hereby granted, free of charge,
 * to any person obtaining a copy of this software
 * and associated documentation files (the "Software"),
 * to deal in the Software without restriction,
 * including without limitation the rights to use, copy,
 * modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software,
 * and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 * 
 * The above copyright notice and this permission notice
 * shall be included in all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
 * THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
 * FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE
 * OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
#include "analyzer.h"
#include <stdio.h>
#include <string.h>

// Static storage for tokenizer function pointers
static tokenizer_ops_t tokenizer_ops;
static int initialized = 0;

// Static arrays for tracking
static int token_type_counts[20];
static char common_words[100][MAX_TOKEN_LENGTH];
static int common_word_counts[100];
static int num_common_words = 0;

void analyzer_init(tokenizer_ops_t ops) {
    tokenizer_ops = ops;
    initialized = 1;
    
    // Reset tracking arrays
    memset(token_type_counts, 0, sizeof(token_type_counts));
    memset(common_word_counts, 0, sizeof(common_word_counts));
    num_common_words = 0;
}

static void track_word(const char *word) {
    // Find if word already exists
    for (int i = 0; i < num_common_words; i++) {
        if (strcmp(common_words[i], word) == 0) {
            common_word_counts[i]++;
            return;
        }
    }
    
    // Add new word
    if (num_common_words < 100) {
        strncpy(common_words[num_common_words], word, MAX_TOKEN_LENGTH - 1);
        common_words[num_common_words][MAX_TOKEN_LENGTH - 1] = '\0';
        common_word_counts[num_common_words] = 1;
        num_common_words++;
    }
}

analysis_result_t analyze_text(const char *text) {
    analysis_result_t result = {0};
    
    if (!initialized) {
        fprintf(stderr, "Error: Analyzer not initialized\n");
        return result;
    }
    
    // Load text using function pointer
    if (tokenizer_ops.load_text(text) != 0) {
        fprintf(stderr, "Error: Failed to load text\n");
        return result;
    }
    
    // Process all tokens using function pointers
    token_t token;
    while ((token = tokenizer_ops.next_token()).type != TOKEN_EOF) {
        // Update counts
        token_type_counts[token.type]++;
        
        switch (token.type) {
            case TOKEN_WORD:
            case TOKEN_IDENTIFIER:
                result.word_count++;
                track_word(token.value);
                break;
                
            case TOKEN_NUMBER:
                result.number_count++;
                break;
                
            case TOKEN_KEYWORD:
                result.keyword_count++;
                break;
                
            case TOKEN_OPERATOR:
                result.operator_count++;
                break;
                
            case TOKEN_COMMENT:
                result.comment_count++;
                break;
                
            case TOKEN_STRING:
                result.string_count++;
                break;
                
            case TOKEN_NEWLINE:
                result.line_count++;
                break;
                
            default:
                break;
        }
    }
    
    // Get final statistics using function pointer
    size_t lines, tokens, chars;
    tokenizer_ops.get_stats(&lines, &tokens, &chars);
    
    result.line_count = lines;
    result.char_count = chars;
    
    return result;
}

void print_token_distribution(void) {
    printf("\n=== Token Distribution ===\n");
    
    const char *token_names[] = {
        "EOF", "WORD", "NUMBER", "PUNCTUATION", "WHITESPACE",
        "NEWLINE", "IDENTIFIER", "KEYWORD", "OPERATOR", 
        "STRING", "COMMENT", "ERROR"
    };
    
    for (int i = 0; i < 12; i++) {
        if (token_type_counts[i] > 0) {
            printf("%s: %d\n", token_names[i], token_type_counts[i]);
        }
    }
    
    printf("\n=== Most Common Words ===\n");
    
    // Simple bubble sort for display
    for (int i = 0; i < num_common_words - 1; i++) {
        for (int j = 0; j < num_common_words - i - 1; j++) {
            if (common_word_counts[j] < common_word_counts[j + 1]) {
                // Swap counts
                int temp_count = common_word_counts[j];
                common_word_counts[j] = common_word_counts[j + 1];
                common_word_counts[j + 1] = temp_count;
                
                // Swap words
                char temp_word[MAX_TOKEN_LENGTH];
                strcpy(temp_word, common_words[j]);
                strcpy(common_words[j], common_words[j + 1]);
                strcpy(common_words[j + 1], temp_word);
            }
        }
    }
    
    // Print top 10
    int limit = num_common_words < 10 ? num_common_words : 10;
    for (int i = 0; i < limit; i++) {
        printf("%d. %s: %d times\n", i + 1, common_words[i], common_word_counts[i]);
    }
}

int calculate_complexity_score(void) {
    int score = 0;
    
    // Base score on keyword density
    score += token_type_counts[TOKEN_KEYWORD] * 2;
    
    // Add points for operators
    score += token_type_counts[TOKEN_OPERATOR];
    
    // Nesting indicators (braces)
    score += token_type_counts[TOKEN_PUNCTUATION] / 10;
    
    // Comments reduce complexity (good documentation)
    score -= token_type_counts[TOKEN_COMMENT];
    
    if (score < 0) score = 0;
    
    return score;
}

void find_patterns(const char *pattern) {
    if (!initialized || !pattern) {
        return;
    }
    
    printf("\n=== Searching for pattern: '%s' ===\n", pattern);
    
    // Reset tokenizer using function pointer
    tokenizer_ops.reset();
    
    int count = 0;
    token_t token;
    
    while ((token = tokenizer_ops.next_token()).type != TOKEN_EOF) {
        if (strstr(token.value, pattern) != NULL) {
            printf("Line %d, Column %d: %s\n", 
                   token.line, token.column, token.value);
            count++;
        }
    }
    
    printf("Found %d occurrences\n", count);
}

File: .cache/claude/global.anthropic.claude-opus-5~5b1m~5d/sha256-c7d2d780d81c14e3adb8f3bc5d0650c7ab1f0fd81df446b75b0c9a994f0fda8a/sha256-e204de8b481288b9491da41674cdb1ef148c487431001035e4e3905ea82a224a/after/translation/src/analyzer.rs
//! Translation of `analyzer.c` / `analyzer.h`.

use std::cell::RefCell;

use crate::cio::{c_strstr, err_str, out_bytes, out_str};
use crate::tokenizer::{TokenType, TokenizerOps, MAX_TOKEN_LENGTH};

#[derive(Default, Copy, Clone)]
pub struct AnalysisResult {
    pub word_count: usize,
    pub number_count: usize,
    pub keyword_count: usize,
    pub operator_count: usize,
    pub comment_count: usize,
    pub string_count: usize,
    pub line_count: usize,
    pub char_count: usize,
}

// Static storage for tokenizer function pointers / static arrays for tracking
struct AnalyzerState {
    tokenizer_ops: Option<TokenizerOps>,
    initialized: bool,
    token_type_counts: [i32; 20],
    common_words: Vec<Vec<u8>>,
    common_word_counts: Vec<i32>,
    num_common_words: i32,
}

impl AnalyzerState {
    fn new() -> AnalyzerState {
        AnalyzerState {
            tokenizer_ops: None,
            initialized: false,
            token_type_counts: [0; 20],
            common_words: vec![Vec::new(); 100],
            common_word_counts: vec![0; 100],
            num_common_words: 0,
        }
    }

    fn track_word(&mut self, word: &[u8]) {
        // Find if word already exists
        for i in 0..self.num_common_words as usize {
            if self.common_words[i] == word {
                self.common_word_counts[i] += 1;
                return;
            }
        }

        // Add new word
        if self.num_common_words < 100 {
            let idx = self.num_common_words as usize;
            let copy_len = word.len().min(MAX_TOKEN_LENGTH - 1);
            self.common_words[idx] = word[..copy_len].to_vec();
            self.common_word_counts[idx] = 1;
            self.num_common_words += 1;
        }
    }
}

thread_local! {
    static STATE: RefCell<AnalyzerState> = RefCell::new(AnalyzerState::new());
}

pub fn analyzer_init(ops: TokenizerOps) {
    STATE.with(|cell| {
        let mut state = cell.borrow_mut();
        state.tokenizer_ops = Some(ops);
        state.initialized = true;

        // Reset tracking arrays
        state.token_type_counts = [0; 20];
        for count in state.common_word_counts.iter_mut() {
            *count = 0;
        }
        state.num_common_words = 0;
    });
}

pub fn analyze_text(text: &[u8]) -> AnalysisResult {
    let mut result = AnalysisResult::default();

    let ops = STATE.with(|cell| {
        let state = cell.borrow();
        if !state.initialized {
            None
        } else {
            state.tokenizer_ops
        }
    });

    let ops = match ops {
        Some(ops) => ops,
        None => {
            err_str("Error: Analyzer not initialized\n");
            return result;
        }
    };

    // Load text using function pointer
    if (ops.load_text)(text) != 0 {
        err_str("Error: Failed to load text\n");
        return result;
    }

    // Process all tokens using function pointers
    loop {
        let token = (ops.next_token)();
        if token.ttype == TokenType::Eof {
            break;
        }

        // Update counts
        STATE.with(|cell| {
            let mut state = cell.borrow_mut();
            state.token_type_counts[token.ttype.index()] += 1;

            match token.ttype {
                TokenType::Word | TokenType::Identifier => {
                    result.word_count += 1;
                    state.track_word(&token.value);
                }

                TokenType::Number => {
                    result.number_count += 1;
                }

                TokenType::Keyword => {
                    result.keyword_count += 1;
                }

                TokenType::Operator => {
                    result.operator_count += 1;
                }

                TokenType::Comment => {
                    result.comment_count += 1;
                }

                TokenType::Str => {
                    result.string_count += 1;
                }

                TokenType::Newline => {
                    result.line_count += 1;
                }

                _ => {}
            }
        });
    }

    // Get final statistics using function pointer
    let (lines, _tokens, chars) = (ops.get_stats)();

    result.line_count = lines;
    result.char_count = chars;

    result
}

pub fn print_token_distribution() {
    out_str("\n=== Token Distribution ===\n");

    const TOKEN_NAMES: [&str; 12] = [
        "EOF",
        "WORD",
        "NUMBER",
        "PUNCTUATION",
        "WHITESPACE",
        "NEWLINE",
        "IDENTIFIER",
        "KEYWORD",
        "OPERATOR",
        "STRING",
        "COMMENT",
        "ERROR",
    ];

    STATE.with(|cell| {
        let mut state = cell.borrow_mut();

        for i in 0..12 {
            if state.token_type_counts[i] > 0 {
                out_str(&format!("{}: {}\n", TOKEN_NAMES[i], state.token_type_counts[i]));
            }
        }

        out_str("\n=== Most Common Words ===\n");

        // Simple bubble sort for display
        let num = state.num_common_words;
        let mut i = 0i32;
        while i < num - 1 {
            let mut j = 0i32;
            while j < num - i - 1 {
                let a = j as usize;
                let b = (j + 1) as usize;
                if state.common_word_counts[a] < state.common_word_counts[b] {
                    // Swap counts
                    state.common_word_counts.swap(a, b);

                    // Swap words
                    state.common_words.swap(a, b);
                }
                j += 1;
            }
            i += 1;
        }

        // Print top 10
        let limit = if state.num_common_words < 10 {
            state.num_common_words
        } else {
            10
        };
        for i in 0..limit {
            let idx = i as usize;
            out_str(&format!("{}. ", i + 1));
            out_bytes(&state.common_words[idx]);
            out_str(&format!(": {} times\n", state.common_word_counts[idx]));
        }
    });
}

pub fn calculate_complexity_score() -> i32 {
    STATE.with(|cell| {
        let state = cell.borrow();
        let mut score: i32 = 0;

        // Base score on keyword density
        score += state.token_type_counts[TokenType::Keyword.index()] * 2;

        // Add points for operators
        score += state.token_type_counts[TokenType::Operator.index()];

        // Nesting indicators (braces)
        score += state.token_type_counts[TokenType::Punctuation.index()] / 10;

        // Comments reduce complexity (good documentation)
        score -= state.token_type_counts[TokenType::Comment.index()];

        if score < 0 {
            score = 0;
        }

        score
    })
}

pub fn find_patterns(pattern: &[u8]) {
    let ops = STATE.with(|cell| {
        let state = cell.borrow();
        if !state.initialized {
            None
        } else {
            state.tokenizer_ops
        }
    });

    let ops = match ops {
        Some(ops) => ops,
        None => return,
    };

    out_str("\n=== Searching for pattern: '");
    out_bytes(pattern);
    out_str("' ===\n");

    // Reset tokenizer using function pointer
    (ops.reset)();

    let mut count: i32 = 0;

    loop {
        let token = (ops.next_token)();
        if token.ttype == TokenType::Eof {
            break;
        }
        if c_strstr(&token.value, pattern) {
            out_str(&format!("Line {}, Column {}: ", token.line, token.column));
            out_bytes(&token.value);
            out_str("\n");
            count += 1;
        }
    }

    out_str(&format!("Found {} occurrences\n", count));
}
